### ISSUE TITLE
Upgrade to terraform 0.13

### ISSUE THREAD

**Issue by digitalronin** [ministryofjustice/cloud-platform#2208]
Terraform 0.13 has been released:
https://www.hashicorp.com/blog/announcing-hashicorp-terraform-0-13

To stay up to date, we need to update all our code to use this version.

This should also allow us to simplify some of our code, but we can do that as we go along. The scope of this ticket is to ensure all our code is compatible with 0.13, and use that as our required terraform version throughout.

**Comment by davidread** [comment]
> We should probably plan & execute our 0.13 upgrade asap, so that we can plan a 0.14 upgrade too

**Comment by AntonyBishop** [comment]
Review and planning upgrade only

**Comment by jasonBirchall** [comment]
https://www.terraform.io/upgrade-guides/0-13.html
https://github.com/hashicorp/terraform/blob/master/CHANGELOG.md

**Comment by jasonBirchall** [comment]
I propose this order of upgrade:
```
cloud-platform-concourse
cloud-platform-terraform-concourse
cloud-platform-terraform-auth0
global-resources

cloud-platform-terraform-awsaccounts-baselines
cloud-platform-terraform-tfstate-backend
cloud-platform-terraform-awsaccounts-iam
cloud-platform-terraform-flow-logs
cloud-platform-account

cross-platform-iam
cloud-platform-network
cloud-platform-eks

cloud-platform-terraform-kops
cloud-platform-terraform-bastion
cloud-platform-terraform-modsec-ingress-controller
cloud-platform-terraform-kops-state-backend
cloud-platform

cloud-platform-terraform-kuberos
cloud-platform-terraform-logging
cloud-platform-terraform-opa
cloud-platform-terraform-monitoring
cloud-platform-terraform-kiam
cloud-platform-terraform-certmanager
cloud-platform-terraform-external-dns
cloud-platform-terraform-ingress-controller
cloud-platform-terraform-starter-pack
cloud-platform-terraform-cluster-autoscaler
cloud-platform-eks-components

cloud-platform-terraform-kube2iam
cloud-platform-terraform-teams-ingress-controller
cloud-platform-dms
cloud-platform-components

cloud-platform-terraform-s3-bucket
cloud-platform-terraform-sqs
cloud-platform-terraform-ecr-credentials
cloud-platform-terraform-elasticache-cluster
cloud-platform-terraform-elasticsearch
cloud-platform-terraform-lambda
cloud-platform-terraform-gitops
cloud-platform-terraform-pushgateway
cloud-platform-terraform-serviceaccount
cloud-platform-terraform-spinaker
cloud-platform-terraform-eventbridge
cloud-platform-terraform-valero
cloud-platform-terraform-sonarqube
cloud-platform-terraform-vault
cloud-platform-terraform-sns-topic
cloud-platform-terraform-rds-aurora
cloud-platform-terraform-rds-instance
cloud-platform-terraform-codepipeline-spike
cloud-platform-terraform-amq-broker
cloud-platform-terraform-dynamo-db
cloud-platform-terraform-rds-snapshot-limit-alert
cloud-platform-terraform-ecr-slack
cloud-platform-terraform-k8s-ingress-controller
cloud-platform-terraform-dynamodb-autoscaler
cloud-platform-environments
```

**Comment by jasonBirchall** [comment]
Before we start this we should also upgrade to Terraform 0.12.30 as it will allow us to ignore versions when performing a `remote_state` lookup.
https://github.com/hashicorp/terraform/blob/v0.12.30/CHANGELOG.md

**Comment by jasonBirchall** [comment]
I have written a script that will walk all of the above repositories and perform a `terraform 013upgrade` command, commit and pr.
https://github.com/jasonBirchall/terraform-update-version/blob/main/main.go

**Comment by mogaal** [comment]
We will have to work on this: 

```
Warning: This resource is deprecated and will be removed in the next major version.                                                                                                                                                                                                                                                                      
Please supply the URL of your repository to helm_release resources directly, using the repository attribute.                                                                                                                                                                                                                                             
See: https://www.terraform.io/docs/providers/helm/r/release.html#example-usage                                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                                                                                         
  on .terraform/modules/cert_manager/main.tf line 45, in data "helm_repository" "jetstack":                                                                                                                                                                                                                                                              
  45: data "helm_repository" "jetstack" {                                                                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                                                                         
(and 9 more similar warnings elsewhere)                                                                                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                                                                                                         
Warning: External references from destroy provisioners are deprecated                                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                                                         
  on .terraform/modules/cert_manager/main.tf line 108, in resource "null_resource" "cert_manager_issuers":                                                                                                                                                                                                                                               
 108:     command = "kubectl delete -n cert-manager -f -<<EOF\n${data.template_file.clusterissuers_production.rendered}\nEOF"                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                                                                                                                         
Destroy-time provisioners and their connection configurations may only                                                                                                                                                                                                                                                                                   
reference attributes of the related resource, via 'self', 'count.index', or                                                                                                                                                                                                                                                                              
'each.key'.                                                                                                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                                                                                                         
References to other resources during the destroy phase can cause dependency                                                                                                                                                                                                                                                                              
cycles and interact poorly with create_before_destroy.                                                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                                                                         
(and 3 more similar warnings elsewhere)                                                                                                                                                                                                                                                                                                                  
```

I am starting to see it with 0.12.30 tf version.

**Comment by jasonBirchall** [comment]
This is also a potential problem. 
https://reports.cloud-platform.service.justice.gov.uk/terraform_modules

Elasticsearch modules are out of sync.

**Comment by jasonBirchall** [comment]
Another issue identified. When upgrading cloud-platform-infrastructure, it directly affects cloud-platform-environments.

**Comment by jasonBirchall** [comment]
https://github.com/ministryofjustice/cloud-platform/issues/2723#issuecomment-797359150

**Comment by jasonBirchall** [comment]
Duplicate ticket https://github.com/ministryofjustice/cloud-platform/issues/2723
